source/memory-store.ts

```diff
@@ -155,69 +167,96 @@ export default class MemoryStore implements Store {
 		void this.resetAll()
 	}
 
-	private resetClient(client: Client, now = Date.now()) {
-		client.totalHits = 0
-		client.resetTime.setTime(now + this.windowMs)
-	}
-
 	/**
-	 * Refill the pool, set previous to current, reset current
+	 * Recycles a client by setting its hit count to zero, and reset time to
+	 * `windowMs` milliseconds from now.
+	 *
+	 * NOT to be confused with `#resetKey()`, which removes a client from both the
+	 * `current` and `previous` maps.
+	 *
+	 * @param client {Client} - The client to recycle.
+	 * @param now {number} - The current time, to which the `windowMs` is added to get the `resetTime` for the client.
+	 *
+	 * @return {Client} - The modified client that was passed in, to allow for chaining.
 	 */
-	private clearExpired() {
-		// At this point, all clients in previous are expired
-
-		// calculate the target pool size
-		this.recentNew.push(this.newClients)
-		if (this.recentNew.length > 10) this.recentNew.shift()
-		this.newClients = 0
-		const targetPoolSize = Math.round(average(this.recentNew))
-
-		// Calculate how many entries to potentially copy to the pool
-		let poolSpace = targetPoolSize - this.pool.length
-
-		// Fill up the pool with expired clients
-		for (const client of this.previous.values()) {
-			if (poolSpace > 0) {
-				this.pool.push(client)
-				poolSpace--
-			} else {
-				break
-			}
-		}
-
-		// Clear all expired clients from previous
-		this.previous.clear()
+	private resetClient(client: Client, now = Date.now()): Client {
+		client.totalHits = 0
+		client.resetTime.setTime(now + this.windowMs)
 
-		// Swap previous and temporary
-		const temporary = this.previous
-		this.previous = this.current
-		this.current = temporary
+		return client
 	}
 
 	/**
-	 * Retrieves or creates a client. Ensures it is in this.current
-	 * @param key IP or other key
-	 * @returns Client
+	 * Retrieves or creates a client, given a key. Also ensures that the client being
+	 * returned is in the `current` map.
+	 *
+	 * @param key {string} - The key under which the client is (or is to be) stored.
+	 *
+	 * @returns {Client} - The requested client.
 	 */
 	private getClient(key: string): Client {
-		if (this.current.has(key)) {
-			return this.current.get(key)!
-		}
+		// If we already have a client for that key in the `current` map, return it.
+		if (this.current.has(key)) return this.current.get(key)!
 
 		let client
 		if (this.previous.has(key)) {
+			// If it's in the `previous` map, return it and bump it up to the `current` map.
 			client = this.previous.get(key)!
```

Comment:
Oh, I think there's a bug here. Client should be *removed* from `previous`, this just gets a reference to it. The client could potentially end up in the `pool` and `current` at the same time, and then get assigned to a new key, thus ending up in `current` more than once and combining the hits from both keys.

I should write a test to make that scenario happen, and then fix the code.

Reply:
Fixed in e0801ce